pkg/deploy/local/local_provider.go

```diff
@@ -0,0 +1,179 @@
+		return err
+	}
+
+	return p.writeConfig(cfg)
+}
+
+// removes the filter from all selected workloads in selected namespaces
+func (p *Provider) RemoveFilter(filter *deploy.Filter) error {
+	cfg, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	if err := removeFilterFromListeners(filter, cfg.GetStaticResources().GetListeners()); err != nil {
+		return err
+	}
+
+	return p.writeConfig(cfg)
+}
+
+// for each hcm in each filter (where it exists)
+func forEachHcm(listeners []*envoy_api_v2.Listener, fn func(networkFilter *envoy_api_v2_listener.Filter, cfg *envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager) error) error {
+	for _, listener := range listeners {
+		for _, chain := range listener.GetFilterChains() {
+			for _, networkFilter := range chain.GetFilters() {
+				if networkFilter.GetName() == util.HTTPConnectionManager {
+					var cfg envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager
+					err := wasmeutil.UnmarshalStruct(networkFilter.GetConfig(), &cfg)
+					if err != nil {
+						return err
+					}
+
+					if err := fn(networkFilter, &cfg); err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func addFilterToListeners(filter *deploy.Filter, listeners []*envoy_api_v2.Listener, filterPath string) error {
+
+	wasmFilter := envoyfilter.MakeWasmFilter(filter, envoyfilter.MakeLocalDatasource(filterPath))
+
+	return forEachHcm(listeners, func(networkFilter *envoy_api_v2_listener.Filter, cfg *envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager) error {
+		for i, httpFilter := range cfg.GetHttpFilters() {
+			if httpFilter.GetName() == wasmeutil.WasmFilterName {
+				// if a wasm filter with the given id exists, return error
+				var wasmFilterConfig config.WasmService
+				err := wasmeutil.UnmarshalStruct(httpFilter.GetConfig(), cfg)
+				if err != nil {
+					return err
+				}
+
+				if wasmFilterConfig.GetConfig().GetName() == filter.ID {
+					return errors.Errorf("filter with id %v already present", filter.ID)
+				}
+			}
+
+			if httpFilter.GetName() == util.Router {
+				// insert the filter before the router
+				cfg.HttpFilters = append(cfg.HttpFilters[:i], wasmFilter, httpFilter)
+
+				// update the HCM with our filter
+				cfgStruct, err := wasmeutil.MarshalStruct(cfg)
+				if err != nil {
+					return err
+				}
+
+				networkFilter.ConfigType = &envoy_api_v2_listener.Filter_Config{
+					Config: cfgStruct,
+				}
+
+				break
+			}
+		}
+		return nil
+	})
+}
+
+func removeFilterFromListeners(filter *deploy.Filter, listeners []*envoy_api_v2.Listener) error {
+	return forEachHcm(listeners, func(networkFilter *envoy_api_v2_listener.Filter, cfg *envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager) error {
+		for i, httpFilter := range cfg.GetHttpFilters() {
+			if httpFilter.GetName() == wasmeutil.WasmFilterName {
+				// if a wasm filter with the given id exists, return error
```

Comment:
is this a todo? Should it be "if a wasm filter with the given id does not exist"

Reply:
i decided to make the behavior "overwrite" instead of error. i'll remove this line